libsolidity/experimental/analysis/TypeInference.cpp

```diff
@@ -0,0 +1,1223 @@
+	{
+		identifierAnnotation.type = handleIdentifierByReferencedDeclaration(_identifierPath.location(), *referencedDeclaration);
+		return false;
+	}
+
+	// TODO: error handling
+	solAssert(false);
+}
+
+bool TypeInference::visit(TypeClassInstantiation const& _typeClassInstantiation)
+{
+	ScopedSaveAndRestore activeInstantiations{m_activeInstantiations, m_activeInstantiations + std::set<TypeClassInstantiation const*>{&_typeClassInstantiation}};
+	// Note: recursion is resolved due to special handling during unification.
+	auto& instantiationAnnotation = annotation(_typeClassInstantiation);
+	if (instantiationAnnotation.type)
+		return false;
+	instantiationAnnotation.type = m_voidType;
+	std::optional<TypeClass> typeClass = std::visit(util::GenericVisitor{
+		[&](ASTPointer<IdentifierPath> _typeClassName) -> std::optional<TypeClass> {
+			if (auto const* typeClassDefinition = dynamic_cast<TypeClassDefinition const*>(_typeClassName->annotation().referencedDeclaration))
+			{
+				// visiting the type class will re-visit this instantiation
+				typeClassDefinition->accept(*this);
+				// TODO: more error handling? Should be covered by the visit above.
+				solAssert(m_analysis.annotation<TypeClassRegistration>(*typeClassDefinition).typeClass.has_value());
+				return m_analysis.annotation<TypeClassRegistration>(*typeClassDefinition).typeClass.value();
+			}
+			else
+			{
+				m_errorReporter.typeError(9817_error, _typeClassInstantiation.typeClass().location(), "Expected type class.");
+				return std::nullopt;
+			}
+		},
+		[&](Token _token) -> std::optional<TypeClass> {
+			if (auto builtinClass = builtinClassFromToken(_token))
+				if (auto typeClass = util::valueOrNullptr(annotation().builtinClasses, *builtinClass))
+					return *typeClass;
+			m_errorReporter.typeError(2658_error, _typeClassInstantiation.location(), "Invalid type class name.");
+			return std::nullopt;
+		}
+	}, _typeClassInstantiation.typeClass().name());
+	if (!typeClass)
+		return false;
+
+	// TODO: _typeClassInstantiation.typeConstructor().accept(*this); ?
+	auto typeConstructor = m_analysis.annotation<TypeRegistration>(_typeClassInstantiation.typeConstructor()).typeConstructor;
+	if (!typeConstructor)
+	{
+		m_errorReporter.typeError(2138_error, _typeClassInstantiation.typeConstructor().location(), "Invalid type constructor.");
+		return false;
+	}
+
+	std::vector<Type> arguments;
+	Arity arity{
+		{},
+		*typeClass
+	};
+
+	{
+		ScopedSaveAndRestore expressionContext{m_expressionContext, ExpressionContext::Type};
+		if (_typeClassInstantiation.argumentSorts())
+		{
+			_typeClassInstantiation.argumentSorts()->accept(*this);
+			auto& argumentSortAnnotation = annotation(*_typeClassInstantiation.argumentSorts());
+			solAssert(argumentSortAnnotation.type);
+			arguments = TypeSystemHelpers{m_typeSystem}.destTupleType(*argumentSortAnnotation.type);
+			arity.argumentSorts = arguments | ranges::views::transform([&](Type _type) {
+				return m_env->sort(_type);
```

Comment:
Is the recursive `TypeEnvrionment::sort()` really what we need here? Why is it recursive in the first place?

For example in something like this:
```solidity
pragma experimental solidity;

type Single(A);
type Double(A, B);

class Self: SomeClass {}
class Self: Number {}
class Self: String {}

instantiation Double(A: Single(T: Number), B: Single(U: String)): SomeClass {}
```

I think this would calculate the sort for `A` as `String` and `B` as `Number`, even though both are clearly stated to be of type `Single` and its their arguments that are constrained.

This example crashes on an assert while visiting `FunctionCall` so something this complex with sorts isn't even fully implemented, but it just isn't the way I think sort should be calculated so maybe there's something I don't understand about them.

Reply:
There's nothing recursive about ``TypeEnvironment::sort`` I can see :-).

Reply:
> There's nothing recursive about `TypeEnvironment::sort` I can see :-).

```c++
		[&](TypeConstant const& _expression) -> Sort
		{
			auto const& constructorInfo = m_typeSystem.constructorInfo(_expression.constructor);
			auto argumentSorts = _expression.arguments | ranges::views::transform([&](Type _argumentType) {
				return sort(resolve(_argumentType));
			}) | ranges::to<std::vector<Sort>>;
			Sort sort;
```

`sort(resolve(_argumentType))` is a recursive call of the `sort()` function.

Reply:
> You can't instantiate like that.

Well, I constructed that by looking at what the parser and the current implementation of `sort()` would handle :) So you're saying that this list will never be nested and we'll only have a flat set of classes there?

```
instantiation Double(A: Class1, B: Class2): SomeClass {}
```

In that case that seems to be all the more reason not to do this recursively.

Reply:
No, that's not what I'm saying :-).
I said you instantiate for a unique type constructor depending on a set of arguments type variables restricted to **sorts**, not classes.
So
```
instantiation Double(A: (Class1, Class2), B: (Class2,Class3)): SomeClass {}
```
is perfectly valid.

That doesn't make things "nested" at all, though.

Reply:
> sort(resolve(_argumentType)) is a recursive call of the sort() function.

So what? How is that an issue if you call it on non-nested input? I really don't see a problem with this :-).

Reply:
Oh, this example finally explains to me why you made this recursive. I did not think that the type constructor here could be a tuple.

> How is that an issue if you call it on non-nested input? I really don't see a problem with this :-).

It is an issue when you're trying to understand what's happening and you don't have much to go off other than the code. I had to have this discussion here with you to figure it out, so I think it's at least missing some explanatory comments (or, if possible, asserts to reject cases that can't happen in practice). Before, I thought it was actually unnecessary, and IMO that would be enough of a reason to make it non-recursive, but now I see that there's a valid case for that. Just, ignoring the constructor without any explanation seemed fishy.

Reply:
> Oh, this example finally explains to me why you made this recursive. I did not think that the type constructor here could be a tuple.

It's a tuple of *classes* (a.k.a. sorts) - it's still a single type constructor that's not a tuple (albeit that it has a tuple of type variables as arguments).

We have to assert that each argument to the one type constructor is just a type variable - but since that may be restricted to sorts, that assertion is easier done *after* looking at the expression. I didn't check again now, but I'm pretty sure that assertion is missing in any case. Or well, it's not an assertion at all actually, it's a case for an error message.